File: src/pocketcloud/cloud/network/packet/impl/normal/PlayerSwitchServerPacket.php
<?php

namespace pocketcloud\cloud\network\packet\impl\normal;

use pocketcloud\cloud\event\impl\player\PlayerSwitchServerEvent;
use pocketcloud\cloud\network\client\ServerClient;
use pocketcloud\cloud\network\Network;
use pocketcloud\cloud\network\packet\CloudPacket;
use pocketcloud\cloud\network\packet\data\PacketData;
use pocketcloud\cloud\player\CloudPlayerManager;
use pocketcloud\cloud\server\CloudServerManager;
use pocketcloud\cloud\terminal\log\CloudLogger;

final class PlayerSwitchServerPacket extends CloudPacket {

    public function __construct(
        private string $playerName = "",
        private string $newServer = ""
    ) {}

    public function encodePayload(PacketData $packetData): void {
        $packetData->write($this->playerName);
        $packetData->write($this->newServer);
    }

    public function decodePayload(PacketData $packetData): void {
        $this->playerName = $packetData->readString();
        $this->newServer = $packetData->readString();
    }

    public function getPlayerName(): string {
        return $this->playerName;
    }

    public function getNewServer(): string {
        return $this->newServer;
    }

    public function handle(ServerClient $client): void {
        if (($player = CloudPlayerManager::getInstance()->get($this->playerName)) !== null) {
            if (($server = CloudServerManager::getInstance()->get($this->newServer)) !== null) {
                Network::getInstance()->broadcastPacket($this);
                CloudLogger::get()->info("Player %s performed a server switch (%s -> %s)", $player->getName(), ($player->getCurrentServer()?->getName() ?? "NULL"), ($server?->getName() ?? "NULL"));
                new PlayerSwitchServerEvent($player, $player->getCurrentServer(), $server)->call();
                $player->setCurrentServer($server);
            }
        }
    }

    public static function create(string $player, string $newServer): self {
        return new self($player, $newServer);
    }
}

File: src/main/java/de/pocketcloud/cloud/network/packet/impl/PlayerSwitchServerPacket.java
package de.pocketcloud.cloud.network.packet.impl;

import de.pocketcloud.cloud.console.log.CloudLogger;
import de.pocketcloud.cloud.event.impl.player.PlayerSwitchServerEvent;
import de.pocketcloud.cloud.network.client.ServerClient;
import de.pocketcloud.cloud.network.packet.CloudboundPacket;
import de.pocketcloud.cloud.network.packet.CloudPacket;
import de.pocketcloud.cloud.network.packet.type.NotificationType;
import de.pocketcloud.cloud.network.packet.data.PacketData;
import de.pocketcloud.cloud.player.CloudPlayerManager;
import de.pocketcloud.cloud.server.CloudServerManager;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.jetbrains.annotations.NotNull;

import java.util.Map;

@NoArgsConstructor
@Getter
public final class PlayerSwitchServerPacket extends CloudPacket implements CloudboundPacket {

    private String player;
    private String newServer;

    public PlayerSwitchServerPacket(String player, String newServer) {
        this.player = player != null ? player : "";
        this.newServer = newServer != null ? newServer : "";
    }

    @Override
    public void handle(@NotNull ServerClient client) {
        var cloudPlayer = CloudPlayerManager.instance().get(player).orElse(null);
        if (cloudPlayer != null) {
            var server = CloudServerManager.instance().get(newServer).orElse(null);
            if (server != null) {
                if (NotificationType.PLAYER_SWITCHED_SERVER.canLog()) {
                    if (cloudPlayer.currentServerName() == null) {
                        CloudLogger.get().info("Player §b{} §rperformed an initial connect on §b{}§r.", cloudPlayer.name(), server.name());
                    } else {
                        CloudLogger.get().info("Player §b{} §rperformed a server switch from §b{} §rto §b{}§r.", cloudPlayer.name(), cloudPlayer.currentServerName(), server.name());
                    }
                }
                NotificationType.PLAYER_SWITCHED_SERVER.notify(Map.of(
                    "player", player,
                    "old_server", cloudPlayer.currentServerName() != null ? cloudPlayer.currentServerName() : "None",
                    "new_server", newServer
                ), Map.of());
                new PlayerSwitchServerEvent(cloudPlayer, cloudPlayer.currentServer().orElse(null), server).call();
                cloudPlayer.setCurrentServer(server);
            }
        }
    }

    @Override
    public void encodePayload(PacketData packetData) {}

    @Override
    public void decodePayload(PacketData packetData) {
        this.player = packetData.readString();
        this.newServer = packetData.readString();
    }

    public static PlayerSwitchServerPacket create(String player, String newServer) {
        return new PlayerSwitchServerPacket(player, newServer);
    }
}
